### DISCUSSION TITLE
OSC control of TA - and problems with ATOM colours!

### DISCUSSION THREAD

**Discussion by Ja5-n** [josephdadams/TallyArbiter#535] [Q&A] [answered]
Hi folks.

First of all thanks for this amazing project, adding tally lights to our hybrid online/in-person performances was wellll beyond the scope of our workflow until we discovered Tally Arbiter!

I'm working with Tally Arbiter 2.0.16 on a Mac Mini M1, and I have a m5 Atom matrix (with tailbat) as our test bed tally light. 

What we're trying to set up is something where I can indicate to a performer which camera they should be focussing on using the tally light. However, this may not necessarily be the active camera! 

So, in this instance I'd like to use OSC commands from qlab or isadora to set the tally light to a red/green state, rather than reading off a program/preview state from e.g. our ATEM. (In actual fact we're using 4 x Sony alpha cameras feeding into a matrix which outputs one 4k signal with each camera's output taking up one 1080 quadrant; this is ingested into both our streaming and production machines so we can use multiple - and different - camera angles in stream and in projection). 

I think I've managed to set up the Devices and Sources correctly: i'm using a Source called CAM1-OSC listening on port 8088, and a Device called CAM1 which is linked to the CAM1-OSC Source.   

When I send an OSC command from Qlab to TA the settings page correctly shows the Device state.
So, sending /tally/program_on "CAM1" results in 
Source: CAM1-OSC Address: CAM1 Label: CAM1 PVW: False PGM: True

BUT! My M5 Atom listener has its colours reversed! When I send CAM1 a PGM instruction the light comes up as Green!

I must be missing something obvious!

Can you help?

J

**Comment by wrldtvlr** [comment, marked as answer]
I recall there was an issue with the red and green colors being reversed in one of the M5 Matrix builds. The problem was that there was common code that didn't take into account that the Matrix was GRB, not RGB. It was fixed in later builds, but as a workaround, you can set custom (inverted) Pgm and Pvw colors.

**Reply by Ja5-n** [reply]
Ah that sounds about right. I mean for the purposes of this current show I can just send the "wrong" osc command, but that seems hella inelegant! Is there anything in the TA UI code for setting custom colours? Or should I be looking at changing the Matrix listener code and re-flashing the Atom? J

**Reply by Ja5-n** [reply]
Looking at the Arduino code the color ints are all labelled as GRB, and the numeric values are set up as GRB - but it looks like the atom is taking the color codes as RGB. So I simply changed the GRB values by swapping the "G" and "R" hex codes around lol

**Reply by Ja5-n** [reply]
So in this snippet I changed the values as described, and didn't change the variable names!

// default color values
int GRB_COLOR_WHITE = 0xffffff;
int GRB_COLOR_BLACK = 0x000000;
//int GRB_COLOR_RED = 0x00ff00; original code
int GRB_COLOR_RED = 0xff0000;
//int GRB_COLOR_ORANGE = 0xa5ff00; original code
int GRB_COLOR_ORANGE = 0xffa500; 
int GRB_COLOR_YELLOW = 0xffff00; // no change G and R values identical
//int GRB_COLOR_GREEN = 0xff0000; original code
int GRB_COLOR_GREEN = 0x00ff00;
int GRB_COLOR_BLUE = 0x0000ff; // no change
//int GRB_COLOR_PURPLE = 0x008080; original code
int GRB_COLOR_PURPLE = 0x800080;
